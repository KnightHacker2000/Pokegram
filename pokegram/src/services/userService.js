/* eslint-disable no-underscore-dangle */
import PokemonClient from '../client';
import API from '../endpoints';
import Activity from '../models/activity';
import User from '../models/user';
import activityService from './activityService';

// const UserService = async () => {
const client = new PokemonClient();
let currSession = null;

/**
 * login API endpoint -- generates random id
 * @param none
*/
const login = async (body) => {
  try {
    const response = await client.post(API.LOGIN, JSON.parse(body));
    currSession = response;
    return response;
  } catch (err) {
    throw new Error(err.response.data);
  }
};

/**
 * logout API endpoint -- clear all session ids
 * @param none
*/
const logout = async () => {
  await client.delete(`${API.LOGIN}/${currSession.id}`);
  // const sessionsIdsArray = (sessions.map((session) => session.id));
  // console.log(sessionsIdsArray);
  // sessionsIdsArray.forEach(async (id) => client.delete(`${API.LOGIN}/${id}`));
  return '200';
};

/**
 * register user API endpoint
 * @param none
*/
const register = async (body) => {
  try {
    const response = await client.post(`${API.USER}`, JSON.parse(body));
    return response;
  } catch (err) {
    throw new Error(err.response.data.error);
  }
};

/**
 * foSug API endpoint
 * @param UID
*/
const getFoSug = async (UID) => {
  // console.log(`${API.SUG}/${UID}`);
  const response = await client.get(`${API.SUG}/${UID}`);
  return response;
};

/**
 * user API endpoint
 * @param userId: user id/ username
*/
const getUserById = async (userId) => {
  // TODO: replace test endpoint
  // const response = await this.client.get(API.USER, userId);
  // console.log(userId);
  const response = await client.get(`${API.USER}/${userId.userId}`);
  const userRes = response.data[0];
  // console.log(userRes);
  const user = new User(
    userRes._id,
    userRes.email,
    userRes.avatar,
    userRes.numPosts,
    userRes.likedPosts,
    userRes.follows,
    userRes.numFollows,
    userRes.subscribers,
    userRes.numSubs,
    userRes.fullname
  );
  // console.log(user);
  return user;
};

const updateUser = async (id, body) => {
  // TODO: replace test endpoint
  // const response = await this.client.get(API.USER, userId);
  // console.log(userId);
  const response = await client.put(`${API.USER}/${id}`, body);
  return response;
};

/**
 * user-like API endpoint
 * @param user: user object
 * @param postId: user liked post
*/
const addlike = async (user, postId, targetUser) => {
  user.likedPosts.push(postId);
  const response = await client.put(`${API.USER}/${user.id}`, user);
  const tmpAct = new Activity();
  tmpAct.initiatorId = user.id;
  tmpAct.targetId = targetUser;
  tmpAct.activityType = 'Like';
  tmpAct.timestamp = Date.now();
  await activityService.createActivity(JSON.stringify(tmpAct));
  return response;
};

/**
 * user-like API endpoint
 * @param user: user object
 * @param postId: user liked post
*/
const removeLike = async (user, postId) => {
  const pIndex = user.likedPosts.indexOf(postId);
  user.likedPosts.splice(pIndex, 1);
  const response = await client.put(`${API.USER}/${user.id}`, user);
  return response;
};

/**
 * user-follow API endpoint
 * @param currentUser: my user object
 * @param targetId: user I followed
*/
const followUser = async (currentUser, targetUser) => {
  const user = currentUser;
  user.follows.push(targetUser.id);
  user.numFollows += 1;
  await client.put(`${API.USER}/${user.id}`, user);

  const newTargetUser = targetUser;
  newTargetUser.subscribers.push(currentUser.id);
  newTargetUser.numSubs += 1;
  await client.put(`${API.USER}/${newTargetUser.id}`, newTargetUser);

  const tmpAct = new Activity();
  tmpAct.initiatorId = currentUser.id;
  tmpAct.targetId = targetUser.id;
  tmpAct.activityType = 'Follow';
  tmpAct.timestamp = Date.now();
  await activityService.createActivity(JSON.stringify(tmpAct));
  // TODO: handle response
  return user;
};

/**
 * user-unfollow API endpoint
 * @param currentUser: my user object
 * @param targetId: user I unfollowed
*/
const unfollowUser = async (currentUser, targetUser) => {
  const user = currentUser;
  const tarIndex = user.follows.indexOf(targetUser.id);
  user.follows.splice(tarIndex, 1);
  user.numFollows -= 1;
  await client.put(`${API.USER}/${user.id}`, user);

  const newTargetUser = targetUser;
  const subIndex = targetUser.subscribers.indexOf(currentUser.id);
  newTargetUser.subscribers.splice(subIndex, 1);
  newTargetUser.numSubs -= 1;
  await client.put(`${API.USER}/${newTargetUser.id}`, targetUser);

  const tmpAct = new Activity();
  tmpAct.initiatorId = currentUser.id;
  tmpAct.targetId = targetUser.id;
  tmpAct.activityType = 'Unfollow';
  tmpAct.timestamp = Date.now();
  await activityService.createActivity(JSON.stringify(tmpAct));
  // TODO: handle response
  return user;
};

export default {
  getUserById,
  login,
  register,
  addlike,
  removeLike,
  followUser,
  unfollowUser,
  logout,
  getFoSug,
  updateUser
};
