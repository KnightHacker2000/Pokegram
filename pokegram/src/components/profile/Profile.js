import PropTypes from 'prop-types';
import React, { useEffect, useState, useRef } from 'react';
import Avatar from '@mui/material/Avatar';
import Button from '@mui/material/Button';
import Box from '@mui/material/Box';
import Container from '@mui/material/Container';
import Divider from '@mui/material/Divider';
import Typography from '@mui/material/Typography';
import { createTheme, ThemeProvider } from '@mui/material/styles';
import userimage from '../../images/jieni.jpg';
import Posts from '../posts/Post';
import userService from '../../services/userService';
import User from '../../models/user';
import HomeState from '../../models/homeState';
// import { TextField } from '@mui/material';

// const user = {
//   id: 1,
//   email: 'Rachel@gmail.com',
//   username: 'Rachel',
//   password: 'RachelYuan',
//   numPosts: 10,
//   likedPost: [1, 2, 3],
//   follows: [1, 2, 3],
//   numFollows: 57,
//   subscribers: [1, 2, 3],
//   numSubs: 1,
//   comments: [1, 2, 3],
//   activities: [1, 2, 3]
// };
const theme = createTheme();
function Profile(props) {
  const { homeStates } = props;
  const firstRendering = useRef(true);
  const [user, setUser] = useState(new User());
  const [myUser, setMyUser] = useState(new User());
  const [isFollow, setIsFollow] = useState(false);
  const [, updateState] = React.useState();
  const forceUpdate = React.useCallback(() => updateState({}), []);

  useEffect(() => {
    // TODO: getUser just testing, use prop.UID in the future
    const params = '{"userId": 2}';
    const myParams = '{"userId": 1}';
    async function fetchData() {
      const data = await userService.getUserById(JSON.parse(params));
      setUser(data);
      await userService.getUserById(JSON.parse(myParams)).then((d) => {
        setMyUser(d);
        if (d.follows.indexOf(2) !== -1) {
          setIsFollow(true);
        }
      });

      // Test code for update endpoint
      // await userService.getUserById(JSON.parse(params)).then((data) => {
      //   setUser(data);
      //   userService.unfollowUser(data, 2);
      // });
    }

    if (firstRendering.current) {
      firstRendering.current = false;
      fetchData();
    }
  }, [isFollow]);

  const handleLogout = (event) => {
    // console.log(props);
    event.preventDefault();
    homeStates.handleHomeStates(
      true,
      homeStates.isProf,
      homeStates.isAct,
      homeStates.isUp,
      homeStates.isPosts,
      homeStates.UID
    );
    // console.log(props);
  };
  const handleFollow = async (event) => {
    event.preventDefault();
    await userService.followUser(myUser, 2);
    firstRendering.current = true;
    setIsFollow(true);
    forceUpdate();
  };

  const handleUnfollow = async (event) => {
    event.preventDefault();
    await userService.unfollowUser(myUser, 2);
    firstRendering.current = true;
    setIsFollow(false);
    forceUpdate();
  };

  function ConditionalRender() {
    let usersection;
    if (homeStates.UID === homeStates.myUID) {
      usersection = (
        <Box sx={{
          mx: 2, p: 1, alignItems: 'center', display: 'flex'
        }}
        >
          <Typography
            component="h1"
            variant="h5"
            sx={{
              marginLeft: '16px',
              marginRight: '64px',
              fontWeight: 'bold',
              font: ''
            }}
          >
            {user.username}
          </Typography>
          <Button type="submit" variant="contained" sx={{ marginTop: '16px', mat: 3, mb: 2 }} onClick={handleLogout}>Log Out</Button>
        </Box>
      );
    } else {
      usersection = (
        <Box sx={{
          mx: 2, p: 1, alignItems: 'center', display: 'flex'
        }}
        >
          <Typography
            component="h1"
            variant="h5"
            sx={{
              marginLeft: '16px',
              marginRight: '64px',
              fontWeight: 'bold',
              font: ''
            }}
          >
            {user.username}
          </Typography>
          {!isFollow && <Button type="submit" variant="contained" sx={{ marginTop: '16px', mat: 3, mb: 2 }} onClick={handleFollow}>Follow</Button>}
          {isFollow && <Button type="submit" variant="contained" sx={{ marginTop: '16px', mat: 3, mb: 2 }} onClick={handleUnfollow}>Unfollow</Button>}
        </Box>
      );
    }
    return (usersection);
  }
  return (
    <ThemeProvider theme={theme}>
      <Container maxWidth="lg">
        <Box
          sx={{
            my: 3, mx: 2, p: 1, marginTop: 0, alignItems: 'center', display: 'flex'
          }}
        >
          <Avatar alt="pikachu" sx={{ height: '100px', width: '100px' }} src={userimage} />
          <Box
            sx={{
              mx: 2, p: 1, marginTop: 5, alignItems: 'center'
            }}
          >
            {ConditionalRender()}
            <Box
              sx={{
                my: 3, mx: 2, alignItems: 'center', display: 'flex'
              }}
            >
              <Typography sx={{ marginLeft: 5, marginRight: 10 }}>{user.numPosts} Posts</Typography>
              <Typography sx={{ marginRight: 10 }}>{user.numSubs} Subscribers</Typography>
              <Typography sx={{ marginRight: 10 }}>{user.numFollows} Follows</Typography>
            </Box>
            <Typography sx={{ marginTop: 0, marginLeft: 7 }}>This is self-Introduction</Typography>
          </Box>
        </Box>
      </Container>
      <Divider />
      <Posts homeStates={homeStates} />
    </ThemeProvider>
  );
}
Profile.propTypes = {
  homeStates: PropTypes.instanceOf(HomeState)
};

Profile.defaultProps = {
  homeStates: null
};
export default Profile;
