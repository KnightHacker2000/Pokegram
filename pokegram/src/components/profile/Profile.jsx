import PropTypes from 'prop-types';
import React, { useEffect, useState, useRef } from 'react';
import Avatar from '@mui/material/Avatar';
import Button from '@mui/material/Button';
import Box from '@mui/material/Box';
import Container from '@mui/material/Container';
import Divider from '@mui/material/Divider';
import Typography from '@mui/material/Typography';
import { createTheme, ThemeProvider } from '@mui/material/styles';
// import userimage from '../../images/jieni.jpg';
import Posts from '../posts/Post';
import userService from '../../services/userService';
import User from '../../models/user';
import HomeState from '../../models/homeState';
import Follows from './Follows';
// import { TextField } from '@mui/material';

// const user = {
//   id: 1,
//   email: 'Rachel@gmail.com',
//   username: 'Rachel',
//   password: 'RachelYuan',
//   numPosts: 10,
//   likedPost: [1, 2, 3],
//   follows: [1, 2, 3],
//   numFollows: 57,
//   subscribers: [1, 2, 3],
//   numSubs: 1,
//   comments: [1, 2, 3],
//   activities: [1, 2, 3]
// };
const theme = createTheme();
function Profile(props) {
  const { homeStates } = props;
  const firstRendering = useRef(true);
  const [user, setUser] = useState(new User());
  const [myUser, setMyUser] = useState(new User());
  const [isFollow, setIsFollow] = useState(false);
  const [showFo, setShowFo] = useState(false);
  const [foSug, setFoSug] = useState([]);
  const [, updateState] = React.useState();
  const [childchange, setchildchange] = React.useState(false);
  const forceUpdate = React.useCallback(() => updateState({}), []);
  const handleischange = () => { setchildchange(!childchange); };
  useEffect(() => {
    // TODO: getUser just testing, use prop.UID in the future
    const params = { userId: homeStates.UID };
    const myParams = { userId: homeStates.myUID };
    async function fetchData() {
      const data = await userService.getUserById(params);
      setUser(data);
      await userService.getUserById(myParams).then((d) => {
        setMyUser(d);
        if (d.follows.indexOf(data.id) !== -1) {
          setIsFollow(true);
        }
      });
      await userService.getFoSug(homeStates.UID).then((d) => { setFoSug(d.data); });

      // Test code for update endpoint
      // await userService.getUserById(JSON.parse(params)).then((data) => {
      //   setUser(data);
      //   userService.unfollowUser(data, 2);
      // });
    }

    if (user.id !== homeStates.UID || firstRendering.current || childchange) {
      firstRendering.current = false;
      setchildchange(false);
      fetchData();
    }
  }, [isFollow, childchange, homeStates]);

  const handleLogout = (event) => {
    // console.log(props);
    event.preventDefault();
    sessionStorage.removeItem('app-token');
    sessionStorage.setItem('logout', true);
    sessionStorage.removeItem('uid');
    homeStates.handleHomeStates(
      true,
      homeStates.isProf,
      homeStates.isAct,
      homeStates.isUp,
      homeStates.isPosts,
      homeStates.UID
    );
    // console.log(props);
  };
  const handleFollow = async (event) => {
    event.preventDefault();
    // console.log(user);
    await userService.followUser(myUser, user);
    firstRendering.current = true;
    setIsFollow(true);
    forceUpdate();
  };

  const handleUnfollow = async (event) => {
    event.preventDefault();
    await userService.unfollowUser(myUser, user);
    firstRendering.current = true;
    setIsFollow(false);
    forceUpdate();
  };

  const handleFollowerCick = () => {
    setShowFo(true);
  };

  const handleShowFo = () => {
    setShowFo(false);
    forceUpdate();
  };

  const getFoList = () => user.follows;

  function ConditionalRender() {
    let usersection;
    if (homeStates.UID === homeStates.myUID) {
      // console.log(user);
      usersection = (
        <Box sx={{
          mx: 2, p: 1, alignItems: 'center', display: 'flex'
        }}
        >
          <Typography
            component="h1"
            variant="h5"
            sx={{
              marginLeft: '16px',
              marginRight: '64px',
              fontWeight: 'bold',
              font: ''
            }}
          >
            {user.id}
          </Typography>
          <Button type="submit" variant="contained" sx={{ marginTop: '16px', mat: 3, mb: 2 }} onClick={handleLogout}>Log Out</Button>
        </Box>
      );
    } else {
      usersection = (
        <Box sx={{
          mx: 2, p: 1, alignItems: 'center', display: 'flex'
        }}
        >
          <Typography
            component="h1"
            variant="h5"
            sx={{
              marginLeft: '16px',
              marginRight: '64px',
              fontWeight: 'bold',
              font: ''
            }}
          >
            {user.id}
          </Typography>
          {!isFollow && <Button type="submit" variant="contained" sx={{ marginTop: '16px', mat: 3, mb: 2 }} onClick={handleFollow}>Follow</Button>}
          {isFollow && <Button type="submit" variant="contained" sx={{ marginTop: '16px', mat: 3, mb: 2 }} onClick={handleUnfollow}>Unfollow</Button>}
        </Box>
      );
    }
    return (usersection);
  }
  return (
    <ThemeProvider theme={theme}>
      { showFo
        && (
        <Follows
          showSug={homeStates.UID === homeStates.myUID}
          foSug={foSug}
          handleShowFo={handleShowFo}
          homeStates={homeStates}
          foList={getFoList()}
        />
        ) }
      <Container maxWidth="lg">
        <Box
          sx={{
            my: 3, mx: 2, p: 1, marginTop: 0, alignItems: 'center', display: 'flex'
          }}
        >
          <Avatar alt="pikachu" sx={{ height: '100px', width: '100px' }} src="http://img4.wikia.nocookie.net/__cb20140328190757/pokemon/images/thumb/2/21/001Bulbasaur.png/200px-001Bulbasaur.png" />
          <Box
            sx={{
              mx: 2, p: 1, marginTop: 5, alignItems: 'center'
            }}
          >
            {ConditionalRender()}
            <Box
              sx={{
                my: 3, mx: 2, alignItems: 'center', display: 'flex'
              }}
            >
              <Typography sx={{ marginLeft: 5, marginRight: 10 }}>
                {user.numPosts}
                Posts
              </Typography>
              <Typography sx={{ marginRight: 10 }}>
                {user.numSubs}
                Subscribers
              </Typography>
              <Typography sx={{ marginRight: 10 }} onClick={handleFollowerCick}>
                {user.numFollows}
                Follows
              </Typography>
            </Box>
            <Typography sx={{ marginTop: 0, marginLeft: 7 }}>This is self-Introduction</Typography>
          </Box>
        </Box>
      </Container>
      <Divider />
      <Posts homeStates={homeStates} ischange={handleischange} />
    </ThemeProvider>
  );
}
Profile.propTypes = {
  homeStates: PropTypes.instanceOf(HomeState)
};

Profile.defaultProps = {
  homeStates: null
};
export default Profile;
