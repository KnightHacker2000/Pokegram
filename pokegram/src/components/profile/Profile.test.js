/**
 * @jest-environment jsdom
 */
import React from 'react';
//  import { fireEvent, render } from '@testing-library/react';
import ShallowRenderer from 'react-test-renderer/shallow';
import Profile from './Profile';
import HomeState from '../../models/homeState';
import '@testing-library/jest-dom/extend-expect';
import { test } from '@jest/globals';
import userService from '../../services/userService';
import { render } from '@testing-library/react';

const parentStates = new HomeState(
  2,
  (isLogOut, newIsProf, newIsAct, newIsUp, newIsPosts, newUID) => {
    console.log(isLogOut, newIsProf, newIsAct, newIsUp, newIsPosts, newUID);
  },
  false,
  false,
  true,
  false,
  1
);

test('test profile snapshot', () => {
  const renderer = new ShallowRenderer();
  renderer.render(<Profile homeStates={parentStates} />);
  const profileComp = renderer.getRenderOutput();
  expect(profileComp).toMatchSnapshot();
});

// test('test profile click', () => {
//   const renderer = new ShallowRenderer();
//   renderer.render(<ResponsiveAppBar homeStates={parentStates} />);
//   const res = renderer.getRenderOutput();
//   // const sha = shallow()
//   // const tree = menuComp.toJSON();
//   res.props.children.props.children.props.children[0].props.onClick();
//   // expect(tree).toMatchSnapshot();
// });

test('test follow button', () => {
  const renderer = new ShallowRenderer();
  renderer.render(<Profile homeStates={parentStates} />);
  const res = renderer.getRenderOutput();
  const spy = jest.spyOn(userService, 'followUser').mockImplementation(() => {});
  res.props.children[1].props.children.props.children[1].props.children[0].props.children[1].props.onClick({ preventDefault: () => {}});
  // res.props.children.props.children.props.children[1].props.children[0].props.children[1].props.onClick({ preventDefault: () => {} });
});