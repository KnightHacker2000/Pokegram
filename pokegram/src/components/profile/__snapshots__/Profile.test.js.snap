// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test profile snapshot 1`] = `
<ThemeProvider
  theme={
    {
      "breakpoints": {
        "between": [Function],
        "down": [Function],
        "keys": [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ],
        "not": [Function],
        "only": [Function],
        "unit": "px",
        "up": [Function],
        "values": {
          "lg": 1200,
          "md": 900,
          "sm": 600,
          "xl": 1536,
          "xs": 0,
        },
      },
      "components": {},
      "direction": "ltr",
      "mixins": {
        "toolbar": {
          "@media (min-width:0px)": {
            "@media (orientation: landscape)": {
              "minHeight": 48,
            },
          },
          "@media (min-width:600px)": {
            "minHeight": 64,
          },
          "minHeight": 56,
        },
      },
      "palette": {
        "action": {
          "activatedOpacity": 0.12,
          "active": "rgba(0, 0, 0, 0.54)",
          "activeChannel": "0 0 0",
          "disabled": "rgba(0, 0, 0, 0.26)",
          "disabledBackground": "rgba(0, 0, 0, 0.12)",
          "disabledOpacity": 0.38,
          "focus": "rgba(0, 0, 0, 0.12)",
          "focusOpacity": 0.12,
          "hover": "rgba(0, 0, 0, 0.04)",
          "hoverOpacity": 0.04,
          "selected": "rgba(0, 0, 0, 0.08)",
          "selectedChannel": "0 0 0",
          "selectedOpacity": 0.08,
        },
        "augmentColor": [Function],
        "background": {
          "default": "#fff",
          "paper": "#fff",
        },
        "common": {
          "black": "#000",
          "white": "#fff",
        },
        "contrastThreshold": 3,
        "divider": "rgba(0, 0, 0, 0.12)",
        "error": {
          "contrastText": "#fff",
          "dark": "#c62828",
          "light": "#ef5350",
          "main": "#d32f2f",
        },
        "getContrastText": [Function],
        "grey": {
          "100": "#f5f5f5",
          "200": "#eeeeee",
          "300": "#e0e0e0",
          "400": "#bdbdbd",
          "50": "#fafafa",
          "500": "#9e9e9e",
          "600": "#757575",
          "700": "#616161",
          "800": "#424242",
          "900": "#212121",
          "A100": "#f5f5f5",
          "A200": "#eeeeee",
          "A400": "#bdbdbd",
          "A700": "#616161",
        },
        "info": {
          "contrastText": "#fff",
          "dark": "#01579b",
          "light": "#03a9f4",
          "main": "#0288d1",
        },
        "mode": "light",
        "primary": {
          "contrastText": "#fff",
          "dark": "#1565c0",
          "light": "#42a5f5",
          "main": "#1976d2",
        },
        "secondary": {
          "contrastText": "#fff",
          "dark": "#7b1fa2",
          "light": "#ba68c8",
          "main": "#9c27b0",
        },
        "success": {
          "contrastText": "#fff",
          "dark": "#1b5e20",
          "light": "#4caf50",
          "main": "#2e7d32",
        },
        "text": {
          "disabled": "rgba(0, 0, 0, 0.38)",
          "primary": "rgba(0, 0, 0, 0.87)",
          "primaryChannel": "0 0 0",
          "secondary": "rgba(0, 0, 0, 0.6)",
          "secondaryChannel": "0 0 0",
        },
        "tonalOffset": 0.2,
        "warning": {
          "contrastText": "#fff",
          "dark": "#e65100",
          "light": "#ff9800",
          "main": "#ed6c02",
        },
      },
      "shadows": [
        "none",
        "0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12)",
        "0px 3px 1px -2px rgba(0,0,0,0.2),0px 2px 2px 0px rgba(0,0,0,0.14),0px 1px 5px 0px rgba(0,0,0,0.12)",
        "0px 3px 3px -2px rgba(0,0,0,0.2),0px 3px 4px 0px rgba(0,0,0,0.14),0px 1px 8px 0px rgba(0,0,0,0.12)",
        "0px 2px 4px -1px rgba(0,0,0,0.2),0px 4px 5px 0px rgba(0,0,0,0.14),0px 1px 10px 0px rgba(0,0,0,0.12)",
        "0px 3px 5px -1px rgba(0,0,0,0.2),0px 5px 8px 0px rgba(0,0,0,0.14),0px 1px 14px 0px rgba(0,0,0,0.12)",
        "0px 3px 5px -1px rgba(0,0,0,0.2),0px 6px 10px 0px rgba(0,0,0,0.14),0px 1px 18px 0px rgba(0,0,0,0.12)",
        "0px 4px 5px -2px rgba(0,0,0,0.2),0px 7px 10px 1px rgba(0,0,0,0.14),0px 2px 16px 1px rgba(0,0,0,0.12)",
        "0px 5px 5px -3px rgba(0,0,0,0.2),0px 8px 10px 1px rgba(0,0,0,0.14),0px 3px 14px 2px rgba(0,0,0,0.12)",
        "0px 5px 6px -3px rgba(0,0,0,0.2),0px 9px 12px 1px rgba(0,0,0,0.14),0px 3px 16px 2px rgba(0,0,0,0.12)",
        "0px 6px 6px -3px rgba(0,0,0,0.2),0px 10px 14px 1px rgba(0,0,0,0.14),0px 4px 18px 3px rgba(0,0,0,0.12)",
        "0px 6px 7px -4px rgba(0,0,0,0.2),0px 11px 15px 1px rgba(0,0,0,0.14),0px 4px 20px 3px rgba(0,0,0,0.12)",
        "0px 7px 8px -4px rgba(0,0,0,0.2),0px 12px 17px 2px rgba(0,0,0,0.14),0px 5px 22px 4px rgba(0,0,0,0.12)",
        "0px 7px 8px -4px rgba(0,0,0,0.2),0px 13px 19px 2px rgba(0,0,0,0.14),0px 5px 24px 4px rgba(0,0,0,0.12)",
        "0px 7px 9px -4px rgba(0,0,0,0.2),0px 14px 21px 2px rgba(0,0,0,0.14),0px 5px 26px 4px rgba(0,0,0,0.12)",
        "0px 8px 9px -5px rgba(0,0,0,0.2),0px 15px 22px 2px rgba(0,0,0,0.14),0px 6px 28px 5px rgba(0,0,0,0.12)",
        "0px 8px 10px -5px rgba(0,0,0,0.2),0px 16px 24px 2px rgba(0,0,0,0.14),0px 6px 30px 5px rgba(0,0,0,0.12)",
        "0px 8px 11px -5px rgba(0,0,0,0.2),0px 17px 26px 2px rgba(0,0,0,0.14),0px 6px 32px 5px rgba(0,0,0,0.12)",
        "0px 9px 11px -5px rgba(0,0,0,0.2),0px 18px 28px 2px rgba(0,0,0,0.14),0px 7px 34px 6px rgba(0,0,0,0.12)",
        "0px 9px 12px -6px rgba(0,0,0,0.2),0px 19px 29px 2px rgba(0,0,0,0.14),0px 7px 36px 6px rgba(0,0,0,0.12)",
        "0px 10px 13px -6px rgba(0,0,0,0.2),0px 20px 31px 3px rgba(0,0,0,0.14),0px 8px 38px 7px rgba(0,0,0,0.12)",
        "0px 10px 13px -6px rgba(0,0,0,0.2),0px 21px 33px 3px rgba(0,0,0,0.14),0px 8px 40px 7px rgba(0,0,0,0.12)",
        "0px 10px 14px -6px rgba(0,0,0,0.2),0px 22px 35px 3px rgba(0,0,0,0.14),0px 8px 42px 7px rgba(0,0,0,0.12)",
        "0px 11px 14px -7px rgba(0,0,0,0.2),0px 23px 36px 3px rgba(0,0,0,0.14),0px 9px 44px 8px rgba(0,0,0,0.12)",
        "0px 11px 15px -7px rgba(0,0,0,0.2),0px 24px 38px 3px rgba(0,0,0,0.14),0px 9px 46px 8px rgba(0,0,0,0.12)",
      ],
      "shape": {
        "borderRadius": 4,
      },
      "spacing": [Function],
      "transitions": {
        "create": [Function],
        "duration": {
          "complex": 375,
          "enteringScreen": 225,
          "leavingScreen": 195,
          "short": 250,
          "shorter": 200,
          "shortest": 150,
          "standard": 300,
        },
        "easing": {
          "easeIn": "cubic-bezier(0.4, 0, 1, 1)",
          "easeInOut": "cubic-bezier(0.4, 0, 0.2, 1)",
          "easeOut": "cubic-bezier(0.0, 0, 0.2, 1)",
          "sharp": "cubic-bezier(0.4, 0, 0.6, 1)",
        },
        "getAutoHeightDuration": [Function],
      },
      "typography": {
        "body1": {
          "fontFamily": ""Roboto", "Helvetica", "Arial", sans-serif",
          "fontSize": "1rem",
          "fontWeight": 400,
          "letterSpacing": "0.00938em",
          "lineHeight": 1.5,
        },
        "body2": {
          "fontFamily": ""Roboto", "Helvetica", "Arial", sans-serif",
          "fontSize": "0.875rem",
          "fontWeight": 400,
          "letterSpacing": "0.01071em",
          "lineHeight": 1.43,
        },
        "button": {
          "fontFamily": ""Roboto", "Helvetica", "Arial", sans-serif",
          "fontSize": "0.875rem",
          "fontWeight": 500,
          "letterSpacing": "0.02857em",
          "lineHeight": 1.75,
          "textTransform": "uppercase",
        },
        "caption": {
          "fontFamily": ""Roboto", "Helvetica", "Arial", sans-serif",
          "fontSize": "0.75rem",
          "fontWeight": 400,
          "letterSpacing": "0.03333em",
          "lineHeight": 1.66,
        },
        "fontFamily": ""Roboto", "Helvetica", "Arial", sans-serif",
        "fontSize": 14,
        "fontWeightBold": 700,
        "fontWeightLight": 300,
        "fontWeightMedium": 500,
        "fontWeightRegular": 400,
        "h1": {
          "fontFamily": ""Roboto", "Helvetica", "Arial", sans-serif",
          "fontSize": "6rem",
          "fontWeight": 300,
          "letterSpacing": "-0.01562em",
          "lineHeight": 1.167,
        },
        "h2": {
          "fontFamily": ""Roboto", "Helvetica", "Arial", sans-serif",
          "fontSize": "3.75rem",
          "fontWeight": 300,
          "letterSpacing": "-0.00833em",
          "lineHeight": 1.2,
        },
        "h3": {
          "fontFamily": ""Roboto", "Helvetica", "Arial", sans-serif",
          "fontSize": "3rem",
          "fontWeight": 400,
          "letterSpacing": "0em",
          "lineHeight": 1.167,
        },
        "h4": {
          "fontFamily": ""Roboto", "Helvetica", "Arial", sans-serif",
          "fontSize": "2.125rem",
          "fontWeight": 400,
          "letterSpacing": "0.00735em",
          "lineHeight": 1.235,
        },
        "h5": {
          "fontFamily": ""Roboto", "Helvetica", "Arial", sans-serif",
          "fontSize": "1.5rem",
          "fontWeight": 400,
          "letterSpacing": "0em",
          "lineHeight": 1.334,
        },
        "h6": {
          "fontFamily": ""Roboto", "Helvetica", "Arial", sans-serif",
          "fontSize": "1.25rem",
          "fontWeight": 500,
          "letterSpacing": "0.0075em",
          "lineHeight": 1.6,
        },
        "htmlFontSize": 16,
        "overline": {
          "fontFamily": ""Roboto", "Helvetica", "Arial", sans-serif",
          "fontSize": "0.75rem",
          "fontWeight": 400,
          "letterSpacing": "0.08333em",
          "lineHeight": 2.66,
          "textTransform": "uppercase",
        },
        "pxToRem": [Function],
        "subtitle1": {
          "fontFamily": ""Roboto", "Helvetica", "Arial", sans-serif",
          "fontSize": "1rem",
          "fontWeight": 400,
          "letterSpacing": "0.00938em",
          "lineHeight": 1.75,
        },
        "subtitle2": {
          "fontFamily": ""Roboto", "Helvetica", "Arial", sans-serif",
          "fontSize": "0.875rem",
          "fontWeight": 500,
          "letterSpacing": "0.00714em",
          "lineHeight": 1.57,
        },
      },
      "zIndex": {
        "appBar": 1100,
        "drawer": 1200,
        "fab": 1050,
        "mobileStepper": 1000,
        "modal": 1300,
        "snackbar": 1400,
        "speedDial": 1050,
        "tooltip": 1500,
      },
    }
  }
>
  <ForwardRef(Container)
    maxWidth="lg"
  >
    <ForwardRef(Box)
      sx={
        {
          "alignItems": "center",
          "display": "flex",
          "marginTop": 0,
          "mx": 2,
          "my": 3,
          "p": 1,
        }
      }
    >
      <ForwardRef(Avatar)
        alt="pikachu"
        src="http://img4.wikia.nocookie.net/__cb20140328190757/pokemon/images/thumb/2/21/001Bulbasaur.png/200px-001Bulbasaur.png"
        sx={
          {
            "height": "100px",
            "width": "100px",
          }
        }
      />
      <ForwardRef(Box)
        sx={
          {
            "alignItems": "center",
            "marginTop": 5,
            "mx": 2,
            "p": 1,
          }
        }
      >
        <ForwardRef(Box)
          sx={
            {
              "alignItems": "center",
              "display": "flex",
              "mx": 2,
              "p": 1,
            }
          }
        >
          <ForwardRef(Typography)
            component="h1"
            sx={
              {
                "font": "",
                "fontWeight": "bold",
                "marginLeft": "16px",
                "marginRight": "64px",
              }
            }
            variant="h5"
          />
          <ForwardRef(Button)
            onClick={[Function]}
            sx={
              {
                "marginTop": "16px",
                "mat": 3,
                "mb": 2,
              }
            }
            type="submit"
            variant="contained"
          >
            Follow
          </ForwardRef(Button)>
        </ForwardRef(Box)>
        <ForwardRef(Box)
          sx={
            {
              "alignItems": "center",
              "display": "flex",
              "mx": 2,
              "my": 3,
            }
          }
        >
          <ForwardRef(Typography)
            sx={
              {
                "marginLeft": 5,
                "marginRight": 10,
              }
            }
          >
            Posts
          </ForwardRef(Typography)>
          <ForwardRef(Typography)
            sx={
              {
                "marginRight": 10,
              }
            }
          >
            Subscribers
          </ForwardRef(Typography)>
          <ForwardRef(Typography)
            onClick={[Function]}
            sx={
              {
                "marginRight": 10,
              }
            }
          >
            Follows
          </ForwardRef(Typography)>
        </ForwardRef(Box)>
        <ForwardRef(Typography)
          sx={
            {
              "marginLeft": 7,
              "marginTop": 0,
            }
          }
        >
          This is self-Introduction
        </ForwardRef(Typography)>
      </ForwardRef(Box)>
    </ForwardRef(Box)>
  </ForwardRef(Container)>
  <ForwardRef(Divider) />
  <Posts
    homeStates={
      HomeState {
        "UID": 2,
        "handleHomeStates": [Function],
        "isAct": false,
        "isPosts": false,
        "isProf": true,
        "isUp": false,
        "myUID": 1,
      }
    }
  />
</ThemeProvider>
`;
